File: src/main/java/studio/dreamys/prometheus/essential/serial/EssentialCosmeticsFileData.java
package studio.dreamys.prometheus.essential.serial;

import gg.essential.lib.gson.Gson;
import gg.essential.lib.gson.JsonParseException;
import gg.essential.lib.gson.reflect.TypeToken;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import studio.dreamys.prometheus.essential.util.GsonUtil;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.lang.reflect.Type;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.LinkedHashSet;
import java.util.Set;
import java.util.TreeSet;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Backing store for the legacy cosmetics list. Serializes as the bare Set's
 * {@code ["a", "b", "c"]} array via Gson.
 */
public class EssentialCosmeticsFileData {
    private static final @NotNull Logger logger = Logger.getLogger("Prometheus - ECFD");
    private static final Gson GSON = new Gson();
    private static final @NotNull Type SET_TYPE = new TypeToken<LinkedHashSet<String>>() {}.getType();
    private static final File COSMETICS_FILE =
            new File(EssentialCosmeticsManager.PROMETHEUS_ESSENTIAL_FOLDER.toFile(), "cosmetics.json");

    /** Guarded by itself. */
    private final @NotNull Set<@NotNull String> legacyCosmetics;
    private boolean isLegacyCosmeticsDirty;


    private EssentialCosmeticsFileData(@NotNull Set<String> legacyCosmetics) {
        this.legacyCosmetics = legacyCosmetics;
    }

    private static final @NotNull EssentialCosmeticsFileData current;

    static {
        Set<String> loaded = new LinkedHashSet<>();
        try {
            if (!COSMETICS_FILE.exists()) {
                Files.createDirectories(COSMETICS_FILE.getParentFile().toPath());
                try (InputStream in = EssentialCosmeticsFileData.class.getClassLoader()
                        .getResourceAsStream("cosmetics.json")) {
                    if (in == null) throw new IllegalStateException("Bundled cosmetics.json not found!");
                    Files.copy(in, COSMETICS_FILE.toPath());
                }
            }
            String json = new String(Files.readAllBytes(COSMETICS_FILE.toPath()), StandardCharsets.UTF_8);
            Set<String> set = GSON.fromJson(json, SET_TYPE);
            if (set != null) loaded = set;
        } catch (IOException e) {
            throw new RuntimeException(e);
        } catch (JsonParseException e) {
            // done to avoid crashing future launches
            logger.log(Level.SEVERE, "cosmetics.json is corrupt, starting with an empty list", e);
        }
        current = new EssentialCosmeticsFileData(loaded);
    }

    // called by MixinCosmeticsManager
    @Contract(value = "-> new", pure = true)
    public static @NotNull Set<@NotNull String> getCosmetics() {
        synchronized (current.legacyCosmetics) {
            return new LinkedHashSet<>(current.legacyCosmetics);
        }
    }

    /** Merges a cosmetic ID into the list (does not save the list to disk). */
    public static boolean addCosmetic(@NotNull String id) {
        boolean added;
        synchronized (current.legacyCosmetics) {
            added = current.legacyCosmetics.add(id);
            if (added) current.isLegacyCosmeticsDirty = true;
        }
        return added;
    }

    /** @see #addCosmetic(String) */
    public static void addCosmetics(@NotNull String @NotNull ...ids) {
        for (String id : ids) addCosmetic(id);
    }


    private static final Object SAVE_COSMETICS_WRITE_LOCK = new Object();
    /** Dumps the cosmetic list to disk. */
    public static void saveCosmetics() {
        synchronized (SAVE_COSMETICS_WRITE_LOCK) {
            Set<String> sorted;
            synchronized (current.legacyCosmetics) {
                if (!current.isLegacyCosmeticsDirty) return;
                sorted = new TreeSet<>(current.legacyCosmetics); // TreeSet is sorted
                current.isLegacyCosmeticsDirty = false;
            }
            try {
                Path tmp = COSMETICS_FILE.toPath().resolveSibling(COSMETICS_FILE.getName() + ".tmp");
                try {
                    GsonUtil.writePretty(tmp, sorted, SET_TYPE);
                    Files.move(tmp, COSMETICS_FILE.toPath(), StandardCopyOption.REPLACE_EXISTING);
                } finally {
                    Files.deleteIfExists(tmp);
                }
            } catch (IOException e) {
                synchronized (current.legacyCosmetics) {
                    current.isLegacyCosmeticsDirty = true; // write failed; keep the change pending for a later retry
                }
                logger.log(Level.SEVERE, "Failed to save cosmetics list!", e);
            }
        }
    }

    /**
     * Downloads and merges from GitHub in the background.
     */
    public static void downloadCosmeticsList() {
        Thread thread = new Thread(() -> {
            try (Reader reader = new InputStreamReader(
                    URI.create("https://github.com/prometheusreengineering/minecraft-essential/raw/refs/heads/main/src/main/resources/cosmetics.json")
                            .toURL().openStream(), StandardCharsets.UTF_8)) {
                String[] cosmetics = GsonUtil.toStringArray(reader);
                addCosmetics(cosmetics);
                saveCosmetics();
                logger.info("Merged " + cosmetics.length + " cosmetics!");
            } catch (Exception e) {
                logger.log(Level.WARNING, "Failed to merge new cosmetics!", e);
            }
        }, "Prometheus - ECFD - DownloadCosmeticsList");
        thread.setDaemon(true);
        thread.start();
    }
}

File: src/main/java/studio/dreamys/prometheus/essential/serial/EssentialCosmeticsFileData.kt
package studio.dreamys.prometheus.essential.serial

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import studio.dreamys.prometheus.essential.ext.File
import studio.dreamys.prometheus.essential.ext.consume
import studio.dreamys.prometheus.essential.serial.EssentialCosmeticsFileData.Companion.COSMETICS_FILE
import studio.dreamys.prometheus.essential.serial.EssentialCosmeticsFileData.Companion.addCosmetic
import studio.dreamys.prometheus.essential.serial.EssentialCosmeticsManager.PROMETHEUS_ESSENTIAL_FOLDER
import java.io.FileOutputStream
import java.net.URI
import java.util.logging.Logger
import kotlin.concurrent.thread

@JvmInline
@Serializable
value class EssentialCosmeticsFileData internal constructor(
    internal val legacyCosmetics: MutableSet<String> = mutableSetOf()
) {
    /* This is a value class so it serializes to the Set's `["a", "b", "c"]` */
    companion object {
        private val logger = Logger.getLogger("Prometheus - ECFD")
        private val COSMETICS_FILE = File(PROMETHEUS_ESSENTIAL_FOLDER, "cosmetics.json")
        private var current: EssentialCosmeticsFileData = run {
            if (!COSMETICS_FILE.exists()) {
                this::class.java.classLoader.getResourceAsStream("cosmetics.json")!!.copyTo(FileOutputStream(COSMETICS_FILE))
            }
            return@run Json.decodeFromString(COSMETICS_FILE.readText())
        }

        // called by MixinServerCosmeticsPopulatePacketHandler
        @JvmStatic
        fun getCosmetics() = current.legacyCosmetics

        /** Merges a cosmetic ID into the list, dumps it to file */
        fun addCosmetic(id: String)
                = addCosmetic(id, true)
        private fun addCosmetic(id: String, save: Boolean): Boolean {
            if (current.legacyCosmetics.contains(id)) return false
            current.legacyCosmetics += id
            if (save) saveCosmetics()
            return true
        }

        /** @see [addCosmetic] */
        fun addCosmetics(vararg ids: String) {
            for (id in ids)
                addCosmetic(id, false)
            saveCosmetics()
        }

        private fun saveCosmetics() {
            COSMETICS_FILE.writeText(Json.encodeToString(current))
        }

        /**
         * Downloads and merges from GitHub in the background.
         * @see COSMETICS_FILE
         */
        fun downloadCosmeticsList() {
            val t = thread {
                val body: String = URI("https://github.com/prometheusreengineering/minecraft-essential/raw/refs/heads/main/src/main/resources/cosmetics.json")
                    .toURL()
                    .openStream()
                    .consume()
                try {
                    val cosmetics = Json.decodeFromString<EssentialCosmeticsFileData>(body).legacyCosmetics.toTypedArray()
                    addCosmetics(*cosmetics)
                    logger.info("Merged ${cosmetics.size} cosmetics!")
                } catch (e: Exception) {
                    logger.warning("Failed to merge new cosmetics!")
                    e.printStackTrace()
                }
            }
        }
    }
}
